api/opds.py

```diff
@@ -938,8 +921,163 @@ def add_authentication_document_link(self, feed_obj):
+                'shared_collection_hold_info',
+                collection_name=self.collection.name,
+                hold_id=active_hold.id,
+                _external=True)
+            kw = dict(href=url, rel='self')
+            info_link_tag = OPDSFeed.makeelement("link", **kw)
+            info_links.append(info_link_tag)
+        return links + info_links
+
+    def revoke_link(self, active_license_pool, active_loan, active_hold):
+        url = None
+        if active_loan:
+            url = self.url_for(
+                'shared_collection_revoke_loan',
+                collection_name=self.collection.name,
+                loan_id=active_loan.id,
+                _external=True)
+        elif active_hold:
+            url = self.url_for(
+                'shared_collection_revoke_hold',
+                collection_name=self.collection.name,
+                hold_id=active_hold.id,
+                _external=True)
+
+        if url:
+            kw = dict(href=url, rel=OPDSFeed.REVOKE_LOAN_REL)
+            revoke_link_tag = OPDSFeed.makeelement("link", **kw)
+            return revoke_link_tag
+
+    def borrow_link(self, identifier,
+                    borrow_mechanism, fulfillment_mechanisms, active_hold=None):
+        if active_hold:
+            borrow_url = self.url_for(
+                "shared_collection_borrow",
+                collection_name=self.collection.name,
+                hold_id=active_hold.id,
+                _external=True,
+            )
+        else:
+            borrow_url = self.url_for(
+                "shared_collection_borrow",
+                collection_name=self.collection.name,
+                identifier_type=identifier.type,
+                identifier=identifier.identifier,
+                _external=True,
+            )
+        rel = OPDSFeed.BORROW_REL
+        borrow_link = AcquisitionFeed.link(
+            rel=rel, href=borrow_url, type=OPDSFeed.ENTRY_TYPE
+        )
+
+        indirect_acquisitions = []
+        for lpdm in fulfillment_mechanisms:
+            # We have information about one or more delivery
+            # mechanisms that will be available at the point of
+            # fulfillment. To the extent possible, put information
+            # about these mechanisms into the <link> tag as
+            # <opds:indirectAcquisition> tags.
+
+            # These are the formats mentioned in the indirect
+            # acquisition.
+            format_types = AcquisitionFeed.format_types(lpdm.delivery_mechanism)
+
+            # If we can borrow this book, add this delivery mechanism
+            # to the borrow link as an <opds:indirectAcquisition>.
+            if format_types:
+                indirect_acquisition = AcquisitionFeed.indirect_acquisition(
+                    format_types
+                )
+                indirect_acquisitions.append(indirect_acquisition)
+
+        if not indirect_acquisitions:
+            raise UnfulfillableWork()
```

Comment:
I agree that this is the right behavior but I was surprised enough that I had to look it up. Could you add an explanatory comment that we want to cancel the creation of the OPDS entry altogether since there's no way to actually get the book?

Reply:
I still think the comment would be good but you didn't add this code, just moved it during the refactor. I probably wrote it originally.